### DISCUSSION TITLE
11.0.x and virtualization : Event Materialized/Dematerialized replacement ?

### DISCUSSION THREAD

**Discussion by Whiletru3** [AvaloniaUI/Avalonia#15324] [Q&A] [answered]
Hello,

I started the migration from 0.10.x to 11.0.10 of our app.
In the new virtualization system, how can I know when an item is materialized and more important is dematerialized.
I manipulate some objects in memory (image, ...) in my viewmodel, and I need to release those in memory objects when the item s dematerialized/recycled.

The interface IItemContainerGenerator the events (Materialized, Dematerialized and Recycled) are gone in the source (but are still in the documentation btw https://reference.avaloniaui.net/api/Avalonia.Controls.Generators/ItemContainerGenerator/ ).

How can I achieve this in 11.0.10+ ?
Thanks :)

**Comment by thevortexcloud** [comment]
The only way I know how to do this is using the ItemsRepeater. It has specific events for when things are generated and removed. The API is exactly the same as it is in UWP so, the UWP docs are just as relevant for Avalonia.

EG https://learn.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.itemsrepeater.elementprepared?view=winui-2.2

I have used it myself to make a fairly decent image gallery. The problem with ItemsRepeater is it has no support for selection out of the box (you can add it yourself though, but it's more work). There may be other ways but I am not sure.

**Comment by Whiletru3** [comment, marked as answer]
I need the selection so ItemsRepeater is not an option.

I found the new replacement in this PR : https://github.com/AvaloniaUI/Avalonia/pull/10685

Now the events are directly in the ItemsControl :)
